### PR TITLE
fix for issue #117 Reliable transport response RSeq header not supported

### PR DESCRIPTION
**Pull request by bkrok**
Please find a solution for issue #117

### CODE REVIEW

File: src/gov/nist/javax/sip/stack/SIPDialog.java [hunk 1]
```diff
@@ -4409,8 +4409,13 @@ public void setEarlyDialogTimeoutSeconds(int seconds) {
...
         final String responseMethod = response.getCSeqHeader().getMethod();
-        final long responseCSeqNumber = response.getCSeq().getSeqNumber();   
-        boolean isRetransmission = !responsesReceivedInForkingCase.add(statusCode + "/" + responseCSeqNumber + "/" + responseMethod);            
+        final long responseCSeqNumber = response.getCSeq().getSeqNumber();
+        RSeq rseq = (RSeq) response.getHeader(RSeqHeader.NAME);
+        String responseRSeqNumber = "";
+        if(rseq != null) {
+            responseRSeqNumber = "/" + rseq.getSeqNumber();
+        }
+        boolean isRetransmission = !responsesReceivedInForkingCase.add(statusCode + "/" + responseCSeqNumber + "/" + responseMethod + responseRSeqNumber);
```
Comment (jaimecasero) [inline src/gov/nist/javax/sip/stack/SIPDialog.java]: I assume if no RSeq header is present, the logic remains as previously,right?

Have you tested this somehow? Any possibility to provide automated junit test?

Ideally these logic should be targeted only for prov resps (1XX) and hopefully enabled form stack conf flag...

File: src/gov/nist/javax/sip/stack/SIPDialog.java [hunk 2]
```diff
@@ -4409,8 +4409,13 @@ public void setEarlyDialogTimeoutSeconds(int seconds) {
     public void checkRetransmissionForForking(SIPResponse response) {
         final int statusCode = response.getStatusCode();
         final String responseMethod = response.getCSeqHeader().getMethod();
-        final long responseCSeqNumber = response.getCSeq().getSeqNumber();   
-        boolean isRetransmission = !responsesReceivedInForkingCase.add(statusCode + "/" + responseCSeqNumber + "/" + responseMethod);            
+        final long responseCSeqNumber = response.getCSeq().getSeqNumber();
+        RSeq rseq = (RSeq) response.getHeader(RSeqHeader.NAME);
+        String responseRSeqNumber = "";
```
Comment (jaimecasero) [inline src/gov/nist/javax/sip/stack/SIPDialog.java]: cant we just compose the intial string here, and then add RSeq part just if necessary.
`
 String retransCondition =(statusCode + "/" + responseCSeqNumber + "/" + responseMethod)
 if(rseq != null) {
            retransCondition = retransCondition  + "/" + rseq.getSeqNumber();
        }
boolean isRetransmission = !responsesReceivedInForkingCase.add(retransCondition)
`

File: src/gov/nist/javax/sip/stack/SIPDialog.java (hunk 1 above)
Comment (bkrok) [inline src/gov/nist/javax/sip/stack/SIPDialog.java]: Your conclusion is right. If the is no RSeq present (reliable transport is not used) the code behaves previously.
Yes this issue was discovered using service specific unit test and the problem disappeared after the fix was implemented. This was the only test we have.
Regarding the optional configuration within that it won't be needed as the additional logic indeed works only for prov resps when they are required. Making responses filtering or implementing configuration will bring some obsolete processing overhead.

File: src/gov/nist/javax/sip/stack/SIPDialog.java (hunk 2 above)
Comment (bkrok) [inline src/gov/nist/javax/sip/stack/SIPDialog.java]: Yes, this is a better way.

Comment (jaimecasero) [inline src/gov/nist/javax/sip/stack/SIPDialog.java]: have you updated the PR with this approach?

File: src/gov/nist/javax/sip/stack/SIPDialog.java (hunk 1 above)
Comment (jaimecasero) [inline src/gov/nist/javax/sip/stack/SIPDialog.java]: So, that junit test was part of your service testsuite. I was wondering If we could add some test to JAIN-SIP test-suite...

Comment (jaimecasero) [inline src/gov/nist/javax/sip/stack/SIPDialog.java]: additional logic indeed works only for prov resps ->ok, it adds almost <ero overhead for rest,so its ok.

Could you add an small comment about RFC3262 appropiate secton that covers the 1XX retrasmissions?

**Comment by jaimecasero** [comment]
@bkrok  if you need merging please review my latest comments..

File: src/gov/nist/javax/sip/stack/SIPDialog.java [hunk 3]
```diff
@@ -4409,8 +4409,17 @@ public void setEarlyDialogTimeoutSeconds(int seconds) {
...
         final int statusCode = response.getStatusCode();
         final String responseMethod = response.getCSeqHeader().getMethod();
-        final long responseCSeqNumber = response.getCSeq().getSeqNumber();   
-        boolean isRetransmission = !responsesReceivedInForkingCase.add(statusCode + "/" + responseCSeqNumber + "/" + responseMethod);            
+        final long responseCSeqNumber = response.getCSeq().getSeqNumber();
+        /*From RFC3262, 1 Introduction: Each provisional response is given a sequence number, carried in the
+        RSeq header field in the response.  The PRACK messages contain an
+        RAck header field, which indicates the sequence number of the
+        provisional response that is being acknowledged.*/
+        RSeq rseq = (RSeq) response.getHeader(RSeqHeader.NAME);
```
Comment (jaimecasero) [inline src/gov/nist/javax/sip/stack/SIPDialog.java]: this looks much cleaner, and with proper comment. Could to rework the PR so there is just one commit with latest changes?

**Comment by bkrok** [comment]
I have created new commit with all changes in one.
https://github.com/ProIDS/jain-sip/commit/bdec230d884797ee8faaa4af68b0b7f9b2ebbea8

**Comment by jaimecasero** [comment]
@bkrok can you create a PR with that single commit?

**Pull request closed without merging**